### ISSUE TITLE
Unable to upload image with "#" in filename

### ISSUE THREAD

**Issue by kagegrifon** [strapi/strapi#6281]
**System**
- Node.js version: 12
- NPM version: 6
- Strapi version: 3.0.0-beta.20
- Database: postgre
- Operating system: Debian GNU/Linux

```
"dependencies": {
    "strapi": "3.0.0-beta.20",
    "strapi-admin": "3.0.0-beta.20",
    "strapi-connector-bookshelf": "3.0.0-beta.20",
    "strapi-plugin-content-manager": "3.0.0-beta.20",
    "strapi-plugin-content-type-builder": "3.0.0-beta.20",
    "strapi-plugin-email": "3.0.0-beta.20",
    "strapi-plugin-upload": "3.0.0-beta.20",
    "strapi-plugin-users-permissions": "3.0.0-beta.20",
    "strapi-utils": "3.0.0-beta.20",
    "knex": "<0.20.0",
    "sqlite3": "latest",
    "pg": "latest"
  },
```

**Describe the bug**
Can't upload an image when filename include symbol "#". With the symbols (&$@ everything is fine.

**Steps to reproduce the behavior**

1. Go to 'Media Library' in cms admin panel (production type)
2. Click on 'Upload assets'
3. Click on 'Browse files'
4. Choose file which name is including "#" symbol
5. Click "Upload ..."
![image](https://user-images.githubusercontent.com/37623678/82337574-99d27d00-99f4-11ea-8df4-2146fd130de0.png)

6. See that file hasn't preview, can't be downloaded, and generally it cannot be used
![image](https://user-images.githubusercontent.com/37623678/82337765-cb4b4880-99f4-11ea-82f1-a6416b486b7c.png)

**Expected behavior**
The image file is uploaded and can be used.

**Screenshots**
![image](https://user-images.githubusercontent.com/37623678/82334670-f6cc3400-99f0-11ea-98f0-ec535eee481d.png)

**Comment by derrickmehaffy** [comment]
Not that I think it will help here but can you test by upgrading to beta.20.3 (also can you test on a fresh rc.1 project?)

**Comment by Brad-Christie** [comment]
> Not that I think it will help here but can you test by upgrading to beta.20.3 (also can you test on a fresh rc.1 project?)

Just tried it out on 20.3 (docker) and looks like it has the same issue. Best i can tell is it's related to final url. Strapi doesn't replace `#` (which is a url special character). Looks like [this line](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/services/Upload.js#L49) might need updating (but unsure of workflows involved/how koa handled file uploads w/ fileInfo so don't want to jump the gun with a PR). If we're in agreement, though, I'd be happy to make the change.

**Comment by lauriejim** [comment]
Hello! Thank you for reporting this issue.
This is due to the `#` character. URLs for the image contains the file name.
And `#` hash has a meaning in URL so break the thing.
Ping @HichamELBSI about that.

Workaround, please rename your file before the upload.

I'm sure you can submit a PR to add a warning.

**Comment by kagegrifon** [comment]
@derrickmehaffy , @Brad-Christie 
thanks for attention for my issue. Most likely, we will set a restriction on the use of the # character in file names for a while.

@lauriejim 
I am not sure that warning will make life easier for those who face this problem.
Or maybe you meant the documentation?

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy , @Brad-Christie 
> thanks for attention for my issue. Most likely, we will set a restriction on the use of the # character in file names for a while.
> 
> @lauriejim 
> I am not sure that warning will make life easier for those who face this problem.
> Or maybe you meant the documentation?

I believe what was meant on adding some kind of error response where if a file is attempted to be uploaded containing that char Strapi would stop and return an error.

**Comment by tudorilisoi** [comment]
This may be solved by encoding the filename portion of the URL
No need to be draconic about restricting file names, some people rely on those for their own intents and purposes

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy @lauriejim @Brad-Christie
> This may be solved by encoding the filename portion of the URL
> No need to be draconic about restricting file names, some people rely on those for their own intents and purposes
> `encodeURIComponent('#example.png') "%23example.png"`
> So the whole matter boils down on how you build the media URL. It should be, well, URLencoded :)

I don't disagree it should be url-encoded, and that URL is stored in the database so I don't see much of an issue with it.

**Comment by lauriejim** [comment]
Ping @alexandrebodin to have his feedback on this subject.

**Comment by Brad-Christie** [comment]
Just noticed this appears to be resolved w/ fa31b5dfc23b7a821f8d7a40a5e7ad8fa278e173 (commit didn't tag issue). Only thing to note is stored filename now removes the special characters (via slugify), though `name` column in `upload_file` table remains intact.

**Comment by alexandrebodin** [comment]
Hi, Yes this is fixed in the 3.0.1 release we just made. I'm closing the issue. Let us know if you still get some similar errors in the future !
